docs/cookbook/named_entity_recognition.md

```diff
@@ -0,0 +1,269 @@
+
+```bash
+export GROQ_API_KEY=your_api_key_here
+```
+
+## Implementation
+
+We'll implement NER with different levels of complexity: simple and nested entity recognition. Let's start with the simple version:
+
+### Simple NER
+
+```python
+from mirascope.core import groq, prompt_template
+from pydantic import BaseModel, Field
+from typing import List
+
+class SimpleEntity(BaseModel):
+    entity: str = Field(description="The entity found in the text")
+    label: str = Field(description="The label of the entity (e.g., PERSON, ORGANIZATION, LOCATION)")
+
+class SimpleNER(BaseModel):
+    entities: List[SimpleEntity] = Field(description="List of named entities found in the text")
+
+@groq.call(
+    model="llama-3.1-8b-instant",
+    response_model=SimpleNER,
+    json_mode=True,
+    call_params=groq.GroqCallParams(temperature=0.0),
+)
+@prompt_template(
+    """
+    Extract the entities from this text: {text}
+    """
+)
+def simple_ner(text: str): ...
+
+complex_text = """
+Apple Inc., the tech giant founded by Steve Jobs and Steve Wozniak, recently announced a partnership with OpenAI, the artificial intelligence research laboratory consisting of the for-profit corporation OpenAI LP and its parent company, the non-profit OpenAI Inc. This collaboration aims to enhance Siri, Apple's virtual assistant, which competes with Amazon's Alexa and Google Assistant, a product of Alphabet Inc.'s Google division. The joint project will be led by Apple's AI chief John Giannandrea, a former Google executive, and will take place at Apple Park, the company's headquarters in Cupertino, California.
+"""
+print("Simple NER Results:")
+simple_result = simple_ner(complex_text)
+for entity in simple_result.entities:
+    print(f"Entity: {entity.entity}, Label: {entity.label}")
+
+# Output
+"""
+Simple NER Results:
+Entity: Apple Inc., Label: ORGANIZATION
+Entity: Steve Jobs, Label: PERSON
+Entity: Steve Wozniak, Label: PERSON
+Entity: OpenAI, Label: ORGANIZATION
+Entity: OpenAI LP, Label: ORGANIZATION
+Entity: OpenAI Inc., Label: ORGANIZATION
+Entity: Amazon, Label: ORGANIZATION
+Entity: Alexa, Label: PRODUCT
+Entity: Google, Label: ORGANIZATION
+Entity: Alphabet Inc., Label: ORGANIZATION
+Entity: John Giannandrea, Label: PERSON
+Entity: Apple Park, Label: LOCATION
+Entity: Cupertino, Label: LOCATION
+Entity: California, Label: LOCATION
+"""
+```
+
+### Nested NER
+
+Now, let's implement a more sophisticated version that can handle nested entities:
+
+```python
+from typing import Optional
+
+class NestedEntity(BaseModel):
+    entity: str = Field(description="The entity found in the text")
+    label: str = Field(description="The label of the entity (e.g., PERSON, ORGANIZATION, LOCATION)")
+    parent: Optional[str] = Field(description="The parent entity if this entity is nested within another entity", default=None)
+    children: List['NestedEntity'] = Field(default_factory=list, description="Nested entities within this entity")
+
+class ImprovedNER(BaseModel):
```

Comment:
same thing here, can just use `list[NestedEntity]`

Reply:
Again i thought it would be a better representation to keep things clear.